packages/langium/src/serializer/json-serializer.ts

```diff
@@ -17,103 +23,84 @@ export interface JsonSerializer {
      * @param node The `AstNode` to be serialized.
      * @param space Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
      */
-    serialize(node: AstNode, space?: string | number): string
+    serialize(node: AstNode, options?: JsonSerializeOptions): string;
     /**
      * Deserialize (parse) a JSON `string` into an `AstNode`.
      */
-    deserialize(content: string): AstNode
+    deserialize(content: string): AstNode;
+}
+
+interface IntermediateReference {
+    $refText?: string
+    $ref: string
+}
+
+function isIntermediateReference(obj: unknown): obj is IntermediateReference {
+    return typeof obj === 'object' && !!obj && '$ref' in obj;
 }
 
 export class DefaultJsonSerializer implements JsonSerializer {
 
-    private readonly linker: Linker;
-    protected ignoreProperties = ['$container', '$containerProperty', '$containerIndex', '$document', '$cstNode'];
+    protected ignoreProperties = new Set(['$container', '$containerProperty', '$containerIndex', '$document', '$cstNode']);
+    protected readonly astNodeLocator: AstNodeLocator;
 
     constructor(services: LangiumServices) {
-        this.linker = services.references.Linker;
+        this.astNodeLocator = services.workspace.AstNodeLocator;
     }
 
-    serialize(node: AstNode, space?: string | number): string {
-        return JSON.stringify(this.decycle(node, this.ignoreProperties), undefined, space);
+    serialize(node: AstNode, options?: JsonSerializeOptions): string {
+        return JSON.stringify(node, (key, value) => this.replacer(key, value, options?.refText), options?.space);
     }
 
     deserialize(content: string): AstNode {
-        return this.revive(JSON.parse(content));
+        const root = JSON.parse(content);
+        this.linkNode(root, root);
+        return root;
     }
 
-    protected decycle(object: AstNode, ignore: string[]): unknown {
-        const objects = new Set<unknown>(); // Keep references to each unique object
+    protected replacer(key: string, value: unknown, refText?: boolean): unknown {
+        if (this.ignoreProperties.has(key)) {
+            return undefined;
+        } else if (isReference(value)) {
+            const refValue = value.ref;
+            return {
+                $refText: refText ? value.$refText : undefined,
+                $ref: '#' + (refValue && this.astNodeLocator.getAstNodePath(refValue))
```

Comment:
How do we handle unresolved references? The `$ref` property should probably be undefined in this case. Shall we store some information from the `error` property in this case (e.g. into `$error`)?